Add readme for database scripting protocol.

KrishnaKulkarni opened the pull request:
https://www.pivotaltracker.com/story/show/133022145

KrishnaKulkarni committed:
Add readme for database scripting protocol.

KrishnaKulkarni commented:
@healthify/development I'm not sure about the proper protocol for changes to this repo, but right now I'm thinking I woudl like approval from a majority of the dev team to go ahead and merge this change.

dleve123 commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+While developing the script, it is critical to plan our remediation strategy if
+something does go awry.
+
+One useful approach to consider is to wrap the INSERT/UPDATE/DELETE commands within
```
small nitpick: I would enclose "INSERT/UPDATE/DELETE" in back-ticks like `INSERT`, etc to make it clear you are referring to SQL.

dleve123 commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+records.
+
+For UPDATE/DELETE operations, we would want to dump the data of all the records
+we will be interacting with into a Temp Table (or Tables). Our reversal script would then
```
when I read "Temp Table", I think there's a formal thing called a Temp Table in SQL or postgreSQL. Is that the case? If not, should we lowercase to "temp table".

amar47shah replied:
Yes, a "temporary table" / "temp table" is a table that lasts the duration of the session.

We should be explicit that the table that you are describing is not a temp table at all, but an ad hoc table backup, that you intend on manually dropping at some time, hours or days later. You might want to coin a term for it, such as "rescue table".

KrishnaKulkarni replied:
@dleve123 It sort of is a specially defined thing in SQL: https://www.postgresql.org/docs/9.3/static/sql-createtable.html

@amar47shah Is correct, however, so we should actually use different terminology. I like "rescue table". Perhaps I should even include a note about creating those in a different schema (like `rescue_tables.*`) for explicit separation from our normal db tables.

dleve123 commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+
+Running database scripts poses a higher risk than deploying a normal feature. Hence,
+it's important to have a higher bar of Code Review than for normal stories. Make sure
+you solicit Code Review from at least two independent reviewers.
```
Potential suggestion: When discussing the follow up to the database incident that motivated this documentation, we noticed that there are 2 levels of "code review": SQL code review (is the SQL valid – will it execute, do we want a certain type of join for performance considerations, etc.), and then "domain sensibility" code review (can we ever delete surveys, can addresses to associated to more than just resource sites, etc.).

Do you think it would be valuable to add a paragraph to describe the latter more – ensuring that someone with deep domain experience reviews the SQL?

georgebrock replied:
There seems to be an implicit assumption here that the code review is going to catch bugs. In my experience, code review isn't very good at this.

Would it make sense to put more emphasis on the a more stringent QA process instead of a more stringent CR process?

amar47shah replied:
> There seems to be an implicit assumption here that the code review is going to catch bugs. In my experience, code review isn't very good at this.

Also there seems to be an implicit assumption that the reversal script will be less prone to bugs than the forward-change script. That assumption seems shaky to me as well.

KrishnaKulkarni replied:
> Do you think it would be valuable to add a paragraph to describe the latter more – ensuring that someone with deep domain experience reviews the SQL?

Yeah I think i discussing those two dimensions to code review would add value.

> Would it make sense to put more emphasis on the a more stringent QA process instead of a more stringent CR process?

Good question. I suppose the reason I was initially inclined to double the CR-ers rather than the QA-ers was because CR seems to have a more defined terminating point, whereas with QA, it's kind of unclear how long you should go about messing around in the app before you're satisfied. That terminating point made me think we might more explicitly get something like double the value from doubling labor on the task, but that could be wrong. I do think you are correct that many bugs will slip by Code Review but may be caught in QA.

@healthify/development Thoughts on adding more CR vs. more QA.

> Also there seems to be an implicit assumption that the reversal script will be less prone to bugs than the forward-change script. That assumption seems shaky to me as well.

We don't need to assume this for there to be value in a reversal script, in that it can give us an extra avenue toward successful restoration, even though sometimes that avenue is suboptimal as well. 

But I also think there is good reason to believe reversal scripts  would be less bug-prone (even if we don't know they're bug free). That reason is that our forward script has to determine a particular set of records to change. The backward script merely needs to retain the info about which set of records were changed then perform a mass COPY or DELETE.

switzersc replied:
> Thoughts on adding more CR vs. more QA.

All great points. I agree it's hard to know when to stop QAing, especially because it's possible (and probable) that our QA does not closely mimic a normal user's behavior (so it's feasible we totally miss a regular workflow that leads to identifying a bug) and also because our staging data is not regularly synced with production data, so the chances of us catching production data edge-cases decreases every day the two datasets stay diverged. I'm not sure what to do about the former, beyond perhaps we should require that more people do independent and/or paired QA. To mitigate the latter, I want to suggest that we more frequently backup our staging database with a sanitized production backup, or at least do so before testing a database script on staging. Thoughts?

KrishnaKulkarni replied:
I agree that backing up/refreshing our staging database more frequently would be a good thing

dleve123 commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+the backup can be used as the source of truth about what the affected database *should*
+look like.
+
+To QA the script, perform the exact the same workflow on staging as you intend for production.
```
extra "the" -> "perform the exact ~~the~~ same..."

dleve123 commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+a particular database and clicking the "Backups" tab. Take note of which backup
+you just created (the backups are timestamped). In the event that there is a problem,
+the backup can be used as the source of truth about what the affected database *should*
+look like.
```
Do you think it would be important to note that we should be using back-ups from `us-east-1`, instead of `us-west-1`? This is something that we ran into while debugging with Frank.

KrishnaKulkarni replied:
wait remind me why this is true?

dleve123 replied:
I believe there are nightly backups to `us-east-1` and `us-west-1` (backups to west incase east is unreachable), but ad-hoc backups are in east-1. I don't remember the exact details, but do remember there being some sort of gotcha. Could be good to confirm with someone from the Aptible Slack channel?

KrishnaKulkarni replied:
I'll check with them

KrishnaKulkarni replied:
They suggested that using either backup would be fine in terms of the data being accurate, but that using the `us-west-1` database would take longer (presumably because of the physical distance to the servers)

dleve123 reviewed (commented):
@KrishnaKulkarni This generally LGTM. I would add some content about the Aptible backup availability zones and "domain sensibility" code review, but nothing seems wrong!

georgebrock commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+
+For INSERT operations, we would need to retain data about what records ended up
+being INSERT-ed by our script, and then the reversal script would delete those
+records.
```
Where should this record be retained? e.g. is there a shared Dropbox folder? What if it contains PHI?

switzersc replied:
I think the point about PHI is really important. I wonder if we should actually have a dedicated section here on handling PHI within scripting.

georgebrock commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+to provide any warnings or other feedback before you execute the script.
+
+Assuming approval from them, perform the same steps from QA:
+1) Create a backup of the staging database.
```
s/staging/production/

georgebrock commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+operations so that we leverage ActiveRecord validations and callbacks.
+
+While developing the script, it is critical to plan our remediation strategy if
+something does go awry.
```
Should we also talk about logging, so we know if something's gone awry, and so we have some data to base any post-script checks on?

KrishnaKulkarni replied:
To be honest, I'm not up-to-date with the state of our logging, so I'm not sure how you would go about doing this. Perhaps you or @switzersc could fill me in on how we could employ such a tactic.

georgebrock replied:
I wasn't thinking of anything more complex than peppering the script with useful `puts` calls, so that after it's finished it's clear what has happened.

KrishnaKulkarni replied:
Gotcha. I suppose so

KrishnaKulkarni commented:
@aaronagray @switzersc @MartinLue Any thoughts here?

switzersc commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+
+While some scripts only lend themselves to being performed via pure SQL, where it
+is possible, you should leverage ActiveRecord to perform our INSERT/UPDATE/DELETE
+operations so that we leverage ActiveRecord validations and callbacks.
```
+1

switzersc commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+
+One useful approach to consider is to wrap the INSERT/UPDATE/DELETE commands within
+a transaction, and then to assert on certain states of the database after the script
+is complete, rolling back all changes if the assertions fail.
```
thoughts on giving an example here? i think it might be helpful to show how one combines SQL and/or active record calls and database state checks with a transaction.

switzersc commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+
+Typically, console scripts are not code that we want to add to the Healthify app's
+code repo. As a result, to solicit Code Review you can put your code in a Github
+[gist](https://help.github.com/articles/about-gists/) and share the gist's link
```
thoughts on specifying that you should use a "secret" gist?

dleve123 replied:
I don't want to introduce a large change late in the game here, but with that said, I'll propose it anyways.

Why a gist? They are harder to version control and code review I imagine - thoughts on committing to some directory in healthify/healthify, say `db/data_migrations/`?

MartinLue replied:
SQL in console scripts _may_ include PHI, but secret gists are not _private_, and can be seen by anyone in possession of the gist URL. So maybe no gist IMO.

amar47shah replied:
+1 to placing scripts in `db/data_migrations`

or even a whole other private repository.

KrishnaKulkarni replied:
👍  Will change

switzersc commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+confirmed, feel free to re-run the script if that would result in a desirable state for
+staging.
+
+As always (but particularly for database scripts), make sure the QA-er is distinct
```
To expand on my last comment regarding QA, what do we think about requiring multiple people to QA rather than just one?

KrishnaKulkarni replied:
I'm ok with that

switzersc commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+1) Create a backup of the staging database.
+2) Execute the script, confirming that it first performs the steps to ensure
+  reversibility (e.g. creates temp tables of the necessary data).
+3) Sign in to the app and explore its functionality, validating that no data was botched.
```
Should this also be someone who's not the developer, or at least someone who is pairing with the developer?

KrishnaKulkarni replied:
That would be preferable. I'll add a comment

switzersc commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,102 @@
+
+After a week has passed without noticeable incident from the script, it is time to
+ensure cleanup of the databases. This entails (a) deprovisioning any unnecessary backup
+databases that were created and (b) destroying any Temp Tables that were created for
```
To confirm: the backup that was manually created sticks around and does not get overwritten by the daily backup?

KrishnaKulkarni replied:
Correct!

KrishnaKulkarni committed:
Improve database scripting README

*Improve articulation of temp tables/rescue tables
*Add section addressing how PHI should be handled
*Add requirement of having 2 QA-ers
*Improve formatting/fix typos

KrishnaKulkarni commented:
@healthify/development Made revisions. Still looking for two more 👍 s!

KrishnaKulkarni commented on protocol/database-scripting/README.md:
```diff
@@ -10,41 +10,54 @@ Develop the scripting story similarly to any other story. In this case, this wou
+
+**Handling PHI**
+
+To remain HIPAA compliant, when our scripts touch data that is PHI, it is critically
```
@dleve123 Does this seem correct?

dleve123 replied:
Yes, this seems correct.

It's not the format of a file that could be "bad" from a security perspective, it's the location of the information. If the CSV lives on Aptible, that's fine. However, chances that if there's a CSV, it'll be emailed/SSHed around, so better to exclude it in the formal documentation.

switzersc commented:
I agree with the change to use `db/data_migrations` rather than a gist. Other than that, 👍 !

amar47shah commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,117 @@
+
+Further, it's important to be able to execute a complete reversal of your changes
+in the event that there are unexpected failures. Ideally, we would write the commands
+we would need to immediately reverse our changes if necessary (i.e. a reversal script).
```
I can't think of a good way to express why, but "reverse" here is somewhat confusing to me. I think "restore the original data" and "restoration script" might be clearer.

amar47shah commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,117 @@
+Development
+-----------
+
+Develop the scripting story similarly to any other story. In this case, this would
```
Instead of using "would"...let's write this document as an imperative set of steps: "Do this, then this..."

If this is checked in to our guide, we should read it as _instructions_ not as a _proposal_.

amar47shah commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,117 @@
+`rescue_table.*` DB schema). Our reversal script would then perform a COPY of the data
+from the rescue table into the original table (and a deletion of updated records in the
+case of an Update). Although an inferior option, it alternatively often suffices to dump
+the data into a csv in lieu of a rescue table.
```
Why is it an inferior option to use a csv? Because there's the potential for spread of PHI, or because it requires the somewhat awkward COPY step to work with? Any more reasons?

On the other hand, using a CSV reduces the likelihood that we will accumulated cruft in our production database. But then, the special `rescue_table` DB schema is a good place to segregate that cruft.

KrishnaKulkarni commented:
@healthify/development Going to revise to include stuff about `db/data_migrations` and also a note on tests, and then 🚢 !

KrishnaKulkarni committed:
Change suggested location of scripts from gist to db/data_migrations

*Also add note about tests for the scripts and fix grammar typos.

KrishnaKulkarni commented:
Made the changes. Will 🚢  within ~1 hour!

amar47shah commented:
@KrishnaKulkarni I left some comments this AM before the merge that it would be nice to see addressed:
- Present document as a "procedure + rationale." Avoid use of "would" and "could".
- Better terminology than "reversal script". In most cases, this won't be a single script but a "restoration strategy" composed of statements to collect a dump and restore data from the dump.
- Reasons why dumping to a CSV is an inferior strategy. (I agree--but it seems weird there are no reasons given!)

KrishnaKulkarni commented:
@amar47shah Hey thanks for the follow up comments. From a workflow perspective, since this is already merged perhaps you can open an Issue or Pull Request to modify this protocol. 

I would argue, however, that unless you think the edits are especially important, it might not be worth prioritizing re-opening this story in over other stories

amar47shah commented:
Sounds good. I think we can leave the story closed for now.

amar47shah commented on protocol/database-scripting/README.md:
```diff
@@ -0,0 +1,121 @@
+For `UPDATE`/`DELETE` operations, we would want to dump the data of all the records
+we will be interacting with into a "rescue table" (a table of the data which we would keep
+for only around a week before deleting. Such a table would ideally live in a
+`rescue_table.*` DB schema). Our reversal script would then perform a COPY of the data
```
We should use `rescue` instead of `rescue_table` as the name of the schema.

See https://github.com/healthify/data_scripting/pull/1. Tagging @switzersc
